ENH: Add pocketfft sources to numpy for testing, benchmarks, etc.

mreineck opened the pull request:
This pull request was opened at the request of @matthew-brett and is related to issues #11885, #11886 and #10625.

ncgeib commented:
I ran the same scripts as I did for #10625 to compare the accuracy and speed. The accuracy issue for real-valued FFTs is fixed and in general pocketfft is faster than the current implementation in numpy but still slower than scipy's fftpack (not taking the additional twiddle factor creation into account). For prime sizes it is vastly faster due to using an algorithm with a lower complexity.

All seems good to me. Only thing I could think of is that maybe the implementation of pocketfft could be sped up further.

![accuracy_complex](https://user-images.githubusercontent.com/9280511/45152357-0da50580-b1d1-11e8-95e1-edb225cedf33.png)
![accuracy_real](https://user-images.githubusercontent.com/9280511/45152360-10075f80-b1d1-11e8-8aa4-735cc5aa4e72.png)
![speed_complex](https://user-images.githubusercontent.com/9280511/45152361-1269b980-b1d1-11e8-81c1-af947361282d.png)
![speed_real](https://user-images.githubusercontent.com/9280511/45152363-14cc1380-b1d1-11e8-8365-915bb01e50c4.png)

mreineck commented:
Sorry, I'm being stupid ... of course I know why pocketfft is currently slower than scipy!

At the moment, pocketfft is not creating plans beforehand and caching them for re-use. That means that for every repetition of the transform it recomputes the twiddle factors all over again.

This can be fixed of course, but I'll probably need help writing a Python class that caches the pocketfft plans.

matthew-brett commented:
I wonder whether some small Cython classes would help here ...

ncgeib commented:
@mreineck Yeah, sorry if that was not clear. The comparison is not really fair as the timings exclude the twiddle factor calculation _and_ cache creation for numpy and scipy and include the twiddle factor calculation for pocketfft.
Still, I am not sure if this explains the whole runtime difference. numpy's C port of fftpack is also slower than the Fortran version. We would have to measure either the twiddle computation or use a cached interface of pocketfft.

mreineck commented:
The code now does some caching.

If someone could tell me the canonical way to pass a C pointer to Python, I'd be grateful. Currently I'm casting its value to a "long" and feel really dirty about it.

charris commented:
> canonical way to pass a C pointer to Python

`NpyCapsule_*`, see `numpy/core/include/numpy/npy_3kcompat.h` and grep for usage.

rgommers commented:
> I ran the same scripts as I did for #10625 to compare the accuracy and speed.

@ncgeib nice. It may be useful to contribute the accuracy benchmarks as tests and the speed ones as `asv` benchmarks to the various packages, if you have time. I doubt our current tests and benchmarks are as comprehensive.

ncgeib commented:
I re-ran the accuracy tests. The comparison should now be fair (and more or less accurate). We see that pocketfft is still just a little bit slower than scipy's fftpack for most test cases. Maybe there is some overhead that could be reduced?

![speed_complex](https://user-images.githubusercontent.com/9280511/45173475-98edbd80-b208-11e8-92eb-6f21fe5c33cf.png)
![speed_real](https://user-images.githubusercontent.com/9280511/45173482-9b501780-b208-11e8-932d-ccc64d0b7775.png)

@rgommers If I find the time I will do that. The benchmarks basically copy the ones FFTW does. For the meantime, they can be found at https://github.com/ncgeib/benchmark-ffts

rgommers commented:
> @rgommers If I find the time I will do that. The benchmarks basically copy the ones FFTW does. For the meantime, they can be found at https://github.com/ncgeib/benchmark-ffts

oh okay, if you mean an actual copy then never mind - FFTW is GPL licensed so we cannot use that code. standalone repo is still useful though, we can then run it again in case of other PRs with major changes like this one

mreineck commented:
@ncgeib I feel much more comfortable with the new benchmarks! The remaining performance difference between scipy and pocketfft is largest for small transform sizes and almost vanishes for longer ones, so my hope is that we can still gain a little bit by optimizing the wrapper code on the Python side.

rgommers commented:
> Hmmm ... I don't see any locking/unlocking going on in the scipy code when the caches are updated. Could this be the reason for the performance difference? And why is scipy not subject to race conditions?

From https://github.com/scipy/scipy/issues/1419: *... This means that SciPy's FFTs are not threadsafe, and thus the GIL cannot be released. Thus, SciPy has to lock up the interpreter while doing FFTs. ...*

mreineck commented:
Maybe there's our performance overhead.

Also, scipy is doing cache management on the C side, which could also help.

But anyway I think this is a detail we can look into once we are convinced that everything else works as intended.

mreineck commented:
There seem to be unnecessary array copies in fftpack.py (and via copy/paste also in pocketfft.py):

The function `fft` starts with:
```
    a = asarray(a).astype(complex, copy=False)
```
(i.e. potentially there is no copy), while `ifft` starts with:
```
    # The copy may be required for multithreading.
    a = array(a, copy=True, dtype=complex)
```
This looks conspicuously inconsistent...

mreineck commented:
BTW, I think that most of numpy.fft's slowness compared to scipy is explained by the many calls to `_check_size()` in helper.py. This is seriously suboptimal code.

mreineck commented:
After some more measurements I'm convinced that the remaining overhead is caused by cache lookup and management. If I use a primitive cache which stores just a single transform, I get speeds that lie practically on top of the scipy data points everywhere.
Not sure what to make of that...

mreineck commented:
This is what I get with the latest patch. Please keep in mind that this is not directly comparable to earlier performance plots, because they were measured on a different computer!

![speed_complex](https://user-images.githubusercontent.com/898507/45224748-0c500780-b2bb-11e8-9ab5-f4dd273f9d6e.png)
![speed_real](https://user-images.githubusercontent.com/898507/45224750-0c500780-b2bb-11e8-8bbc-06879ebcaabe.png)

ncgeib commented:
I can confirm these on my machine. In the latest state of the PR the runtime of pocketfft is basically the same as scipy's fftpack for regular numbers.

mreineck commented:
Great!
So, what's the procedure from here on?

All Travis builds fail (even those for Python 3), because C89 is enforced. Do we need to request a change in policy here?

charris commented:
I'm thinking of branching 1.16.0 around the end of November.

charris commented:
You can try changing the build/test environment in this PR to see what needs doing, but it would be best to keep those changes as separate commits as we might want to split the resulting PR when we move up. In particular, you can drop the 2.7 tests and look to making the rest of NumPy C99 safe. If some of that can be done without breaking 2.7 on windows, so much the better.

charris commented:
I've put up #11905 to start testing with c99 on travisCI.

mreineck commented:
Thanks a lot for that, @charris ! I'm not familiar with configuration in the Python world and couldn't have done this myself.

charris commented:
You need to use the `NPY_RESTRICT` keyword, as msvc doesn't support it directly.

mreineck commented:
Will try this tomorrow.

charris commented:
Looks like the remaining problems are declarations after code. We can fix that two ways, either in the code or in the tests. Until we officially require C99 I would prefer the first, but you can do the easy fix and remove that error from https://github.com/numpy/numpy/blob/master/tools/travis-test.sh#L28.

charris commented:
Note the `NPY_RESTRICT` is defined to `restrict` for gcc, doing it the other way round defeats the purpose :) You need to make the replacement with an editor and search-and-replace.

mreineck commented:
@charris Well, if NPY_RESTRICT had been defined on the command line, it might have worked, although this kind of circular macro expansion is definitely not what we want ;)
I see that the macro is defined in `numpy/core/include/numpy/config.h`. Can I assume that this file is present when pocketfft is compiled?

BTW, I also noticed that my caching implementation is not thread-safe; I'll rework it, but it will take some time.

charris commented:
> Can I assume that this file is present when pocketfft is compiled?

It is created during build. On my machine,
```
charris@fc [numpy.git (master)]$ grep -r RESTRICT build
...
build/src.linux-x86_64-2.7/numpy/core/include/numpy/config.h:#define NPY_RESTRICT restrict
```

charris commented:
Some of the windows builds are passing now. For linux you still have the declaration after code errors, which is easiest to fix by removing the indicated part of `travis-test.sh`.

mreineck commented:
Yes, looking much better now!
I didn't realize that you could provoke gcc to actually mark declarations after code as errors when in C99 mode.
If possible, I'd really like to stick to C99 for the library, at least if we expect the PR to be merged after dropping Python 2.7. If there was consensus that this should go in earlier, I can make the changes, but the code would definitely not look any nicer afterwards.

charris commented:
Looks like the sdist fails. I think that can be fixed by adding the `pocketfft.inc` to `MANIFEST.in` and some fixups to the `setup.py` adding dependencies and such. Maybe just the latter. Python2.7 fails on windows as expected, we will just have to wait until we drop 2.7 support to fix that.

charris commented:
> I didn't realize that you could provoke gcc to actually mark declarations after code as errors when in C99 mode.

Neither did I, it is a learning experience...

For the rest, looks like we will shoot for full C99, or at least as much as Window allows, it doesn't seem to make much sense not to.

charris commented:
If you want to fixup the rest of the errors, you could also drop Python 2.7 from the test matrix. Or I could put  up a PR to do that with the understanding that it waits on branching 1.16. Or we could just wait and rebase this after the branch.

charris commented:
> Can I assume that this file is present when pocketfft is compiled?

Looking at this more carefully, the file is generated when `numpy/core/setup.py` is run, so it does depend on the order in which things are built. It isn't clear to me that that is safe if we run parallel builds.

mreineck commented:
OK, I have removed fftpack and the branch now uses pocketfft for all purposes.
All tests should pass now. I added a few of my own, overall coverage seems very good.

There should be enough time for polishing this until 1.17 :)

mreineck commented:
I did some more performance tuning, which mostly improves run times of FFTs with prime factors >5. The comparison to scipy's Fortran implementation is shown below. (Benchmarking was done for the C parts only, to be independent of Python overhead.)
![fft_benchmark](https://user-images.githubusercontent.com/898507/45568512-28285000-b85d-11e8-9639-f18be75d8ee9.png)

mreineck commented:
I'm wondering whether we can/should abandon plan caching altogether...
Plan generation in the new implementation is much faster than in the original fftpack: fftpack uses O(n) sin/cos calls for an FFT of length n, pocketfft needs roughly 2*sqrt(n) (the rest is computed using angle addition theorems).
This would mean:
- a performance hit for short (n<100) 1D FFTs
- a performance hit of 50% for large prime-sized 1D FFTS (but remember, they will still be much, much faster than with fftpack)
- no measurable hit for multi-D FFTs (because the plan for the 1D FFTs needed for such a transform would still be re-used).
- quite some simplification on the Python side.

Opinions?

ncgeib commented:
Is there any more reason than "simplification"? If the caching code is already written and it works, this would not be a strong reason then. Do you think it impacts maintainability significantly? Or are there other issues? To be honest though, I did not take a good look at the caching code yet but planned to do it before this PR is merged.

My concerns are:
- I have a use-case (1D pulse propagation using a split-step algorithm) where the main computation time is spent on 1D-FFTs with small n (e.g., n ~= 128).
- The actual running time of the twiddle-factor computation depends on the libmath that is used. Depending on the platform the implementation of sin and cos may be slower than what you have on your system (a recent glibc probably).
- I was thinking of further increasing the accuracy by shipping a custom implementation of sincos(pi*x) (like it is done in the sleef library). Chances are that this actually speeds up the computation - but by not caching the computation it would have to.

charris commented:
> the rest is computed using angle addition theorems.

How accurate is that? I've done similar with the divide and conquer (log2) approach many years ago and ISTR that even in that case one needed to be careful about accuracy.

mreineck commented:
> How accurate is that?

As accurate as the plots above indicate.
This is done in a way that sin/cos of every angle are computed with at most one application of an angle addition theorem from "exact" sin/cos pairs.

The idea is like this: If the 100 sin/cos pairs of `n*2*pi/100` are needed, compute the 10 sin/cos pairs of `n*2*pi/10` and the first 10 sin/cos pairs of `n*2*pi/100`. Everything else is computable with one application of the addition formula.

If you go for log(n) sin/cos evaluation as opposed to sqrt(n), then inaccuracies can become a problem.

mreineck commented:
> Is there any more reason than "simplification"?

Not really. Well, we would get rid of a mutex too, but I don't know whether this is important.

> I have a use-case (1D pulse propagation using a split-step algorithm) where the main computation time is spent on 1D-FFTs with small n (e.g., n ~= 128).

In that case the current caching approach (checking in a dictionary whether the desired length is available, then popping this entry and reinserting it) will probably hurt you more than plan recalculation. (This is just a guess, I haven't measured this concrete case)

> The actual running time of the twiddle-factor computation depends on the libmath that is used. Depending on the platform the implementation of sin and cos may be slower than what you have on your system (a recent glibc probably).

Absolutely, but again this should only matter for short FFTs, but for those the current caching mechanism is already quite slow.

> I was thinking of further increasing the accuracy by shipping a custom implementation of sincos(pi*x) (like it is done in the sleef library). Chances are that this actually speeds up the computation - but by not caching the computation it would have to.

I'm already using an implementation that maps angles to the [0;pi/4] range to achieve the best possible accuracy. If there are any other tricks, I'd be happy if I could have a peek!

rgommers commented:
IIRC there are some open issues with the caching mechanism:
https://github.com/numpy/numpy/issues/7150#issuecomment-177897404
https://github.com/numpy/numpy/issues/1864
There may be more. If performance is okay, then dropping the caching mechanism completely will result in more robust code.

matthew-brett commented:
Is there any way to allow users more easily to implement their own caching?  I guess that will be a reasonably common desire, and it would be good to help reduce code duplication and bugs.

ncgeib commented:
As far as I know the PR currently uses no recursion to calculate the twiddle factors (trig_utils.c, line 146+147, last argument = 1). The accuracy benchmarks were done for this option. Activating the recursion (last argument = 0) is slightly detrimental to the accuracy (see below). If we use caching I would not recommend using it. But of course, if we decide to drop caching it would make sense.

![accuracy_complex](https://user-images.githubusercontent.com/9280511/45895389-5b289180-bdd1-11e8-8386-6197fdcb9c42.png)

> In that case the current caching approach (checking in a dictionary whether the desired length is available, then popping this entry and reinserting it) will probably hurt you more than plan recalculation. (This is just a guess, I haven't measured this concrete case)

Intuitively I would say that retrieving from the cache is slower than than the computation of 2*sqrt(N) transcendental function values, there should be ways to improve it, I guess. But maybe I am wrong.

> I'm already using an implementation that maps angles to the [0;pi/4] range to achieve the best possible accuracy. If there are any other tricks, I'd be happy if I could have a peek!

Sure, contact me if you want to work on that. I spent a lot of time reading up on it and know how it *should* work but never got it finished.
The idea here is that the library functions `sin(x) `and `cos(x)` may be correctly rounded with respect to `x`, but you obviously already lose accuracy when calculating `pi*x.` The solution is to implement a version of `sin(pi*y)` which is correctly rounded with respect to `y`. Also there is an advantage of computing `sin(x)` and `cos(x)` at the same time. Implementing such a function that works for all `x` requires fancy range reduction algorithms. In our use case, however, we only need it to work on `[0, 1/4]` which simplifies this immensely.

It *should* give additional accuracy but I never got to test how much... ;)

mreineck commented:
> As far as I know the PR currently uses no recursion to calculate the twiddle factors (trig_utils.c, line 146+147, last argument = 1).

Yes, this part is computed accurately. The shortcut I'm referring to is in lines 154 and 155, where I'm combining two of those "exact" values to get the result for the requested angle.

> The idea here is that the library functions sin(x)and cos(x) may be correctly rounded with respect to x, but you obviously already lose accuracy when calculating pi*x. The solution is to implement a version of sin(pi*y) which is correctly rounded with respect to y.

Since all sin/cos pairs that are needed in an FFT are nth roots of unity it is possible to do the range reduction already in the integer domain and avoid those rounding problems entirely.
If I need `sin (2*pi*1337/1338)`, I can just compute `sin(2*pi*1/1338)` and use that value.
I need to rely on the CPU to give me the best value for `(2*pi*n/m)`, where `n<=m/8`, but this should be fine.

> Also there is an advantage of computing sin(x) and cos(x) at the same time. 

All compilers I know do this automatically these days. If I forget to link pocketfft to libm, I get error messages about "undefined symbol sincos()", so this is luckily not an issue.

mreineck commented:
@ncgeib Could you please give your pulse propagation code a try with the current branch? Caching is now completely disabled.

ncgeib commented:
> Yes, this part is computed accurately. The shortcut I'm referring to is in lines 154 and 155, where I'm combining two of those "exact" values to get the result for the requested angle.

Ah, I should read the code more carefully next time... Sorry for the noise then.

> Since all sin/cos pairs that are needed in an FFT are nth roots of unity it is possible to do the range reduction already in the integer domain and avoid those rounding problems entirely.
If I need `sin (2*pi*1337/1338)`, I can just compute `sin(2*pi*1/1338)` and use that value.
I need to rely on the CPU to give me the best value for `(2*pi*n/m)`, where `n<=m/8`, but this should be fine.

The range reduction with the integer values is not the point here, it will be used in either case. The point is the multiplication with pi: in `sin(pi*x)` it is performed with a double accuracy approximation of pi, in `sinpi(x)` it is done implicitly inside the function with higher accuracy. The former case leads to a less accurate result. The standards (IEEE and C99) recommend that those functions are available, but most libms do not implement them. One would have to ship them with the fft library - which seems feasible as only an implementation for a very limited range of `x` has to be provided. [sleef](http://sleef.org/additional.xhtml#sincospi) has an implementation, but it is tuned for speed and not for accuracy.

> All compilers I know do this automatically these days. If I forget to link pocketfft to libm, I get error messages about "undefined symbol sincos()", so this is luckily not an issue.

`sincos` is not part of the standard math.h. I know that it is available as a GNU extension in gcc and the replacement you mention will be used on linux. But I fear [this is not the case](https://stackoverflow.com/questions/24328173/c-libstd-compute-sin-and-cos-simultaneously#comment88768981_24331473) using Visual Studio or clang.

> @ncgeib Could you please give your pulse propagation code a try with the current branch? Caching is now completely disabled.

will do, but may take some time.

mreineck commented:
https://stackoverflow.com/questions/42792939/implementation-of-sinpi-and-cospi-using-standard-c-math-library has a beautiful (and not too slow) reference implementation of sincospi. I'll have a closer look, but this is a somewhat orthogonal project. I don't expect any significant effect on the FFT errors.

ncgeib commented:
> https://stackoverflow.com/questions/42792939/implementation-of-sinpi-and-cospi-using-standard-c-math-library has a beautiful (and not too slow) reference implementation of sincospi. I'll have a closer look, but this is a somewhat orthogonal project. I don't expect any significant effect on the FFT errors.

Any further improvement of the FFT errors will necessarily be very marginal (if it exists at all). The current implementation is almost limited by the accuracy of the additions and multiplications and not by the accuracy of the twiddle-factors. However, if one is willing to spend the extra effort it would be possible to use correctly rounded roots of unity by, e.g., a custom `sincospi` function. If implemented correctly it could even improve the performance of the twiddle-factor computation and would make the accuracy portable (currently it depends on the libm). I totally agree that this is an orthogonal project and simply something that I was interested in.

The accuracy impact of such a change can be tested by calculating the twiddle-factors in extended precision (which I observed to have a rather larger perfomance impact). This should give faithfully rounded twiddle-factors, the same as the implementation you linked to (<1 ulp whereas correctly rounded is <0.5 ulp). When I did it for #10625 I noticed only a very small accuracy improvement.

mreineck commented:
Current results without caching:
![accuracy_complex](https://user-images.githubusercontent.com/898507/45920775-b928a800-bea9-11e8-813a-2bde0eebd3cf.png)
![accuracy_real](https://user-images.githubusercontent.com/898507/45920776-b928a800-bea9-11e8-90ac-8b9eea7fb5ed.png)
![speed_complex](https://user-images.githubusercontent.com/898507/45920777-b928a800-bea9-11e8-9d80-d385e70d9b43.png)
![speed_real](https://user-images.githubusercontent.com/898507/45920778-b9c13e80-bea9-11e8-8ca9-45cea602bc94.png)

rgommers commented:
nice @mreineck. conclusion from those graphs is that removing the caching mechanism completely is okay, correct?

mreineck commented:
From a pragmatic point of view, I'd say yes. Some transforms are noticeably slower than they could be (especially the Bluestein ones)... but everyone _really_ concerned about performance will use FFTW anyway.

Another question is of course whether the code in this form could still be a candidate for replacing scipy's FFT. But maybe now is not the time to worry about this.

rgommers commented:
> Another question is of course whether the code in this form could still be a candidate for replacing scipy's FFT.

I think so. In most cases it's competitive, and it does have the Bluestein algorithm which SciPy lacks.

> But maybe now is not the time to worry about this.

Agreed.

ncgeib commented:
Trying out the current PR I see the same behavior in my use-cases. Dropping the caching does not lead to significant slow down for common choices of `N` (i.e., non-prime) - quite the contrary.

I severely underestimated the overhead of the Python side. I did a [line profiling](https://github.com/rkern/line_profiler) for 100 000 calls of a complex-valued FFT with `N=1024` of the last version with cache (commit cefa941):

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    51                                               def get_plan(self, n):
    52    100000      82686.0      0.8     32.5          with self._lock:
    53    100000      37093.0      0.4     14.6              pos = self._pos
    54    100000      53128.0      0.5     20.9              if self._n[pos] == n:
    55     99999      81388.0      0.8     32.0                  return self._plans[pos]
    62         1         30.0     30.0      0.0              self._plans[self._pos] = self._constr(n)

    70                                           def _raw_fft(a, n, axis, cache, is_real, is_forward, fct):
    71    100000     137179.0      1.4      5.3      a = asarray(a)
    73    100000      59132.0      0.6      2.3      if n is None:
    74                                                   n = a.shape[axis]
    76    100000      54425.0      0.5      2.1      if n < 1:
    80    100000      88604.0      0.9      3.4      if a.shape[axis] != n:
    94    100000     667427.0      6.7     25.7      plan = cache.get_plan(n)
    95    100000      65427.0      0.7      2.5      if axis == -1 or axis == a.ndim-1:
    96    100000    1452777.0     14.5     56.0          r = pfi.execute(a, plan, is_real, is_forward, fct)
   101    100000      70038.0      0.7      2.7      return r

   113                                           def fft(a, n=None, axis=-1, norm=None):
   200    100000     335934.0      3.4      8.3      a = asarray(a).astype(complex, copy=False)
   201    100000      60375.0      0.6      1.5      if n is None:
   202    100000      80856.0      0.8      2.0          n = a.shape[axis]
   203    100000      49357.0      0.5      1.2      if norm is None:
   204    100000      47438.0      0.5      1.2          fct = 1.
   207    100000    3425701.0     34.3     84.7      output = _raw_fft(a, n, axis, _CCache, False, True, fct)
   208    100000      46252.0      0.5      1.1      return output
```
Time unit is µs. I omitted the lines which do not contribute or are never executed. Just obtaining the plan from the cache costs 6.7µs compared to the 14.5µs for executing the whole FFT. A substantial part of this time is required by the lock.

Comparing this to the current state of the PR without caching we have:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    40                                           def _raw_fft(a, n, axis, is_real, is_forward, fct):
    41    100000     158806.0      1.6      7.2      a = asarray(a)
    43    100000      63423.0      0.6      2.9      if n is None:
    46    100000      57295.0      0.6      2.6      if n < 1:
    50    100000      84752.0      0.8      3.8      if a.shape[axis] != n:
    64    100000      60282.0      0.6      2.7      if axis == -1 or axis == a.ndim-1:
    65    100000    1707124.0     17.1     77.2          r = pfi.execute(a, is_real, is_forward, fct)
    70    100000      79729.0      0.8      3.6      return r
```
The non-cached call for the FFT takes 17.1µs - clearly the difference is smaller than the overhead of the cache...

I guess that caching the twiddle-factors would only make sense in a very efficient Cython wrapper (or when caching on C-side). Maybe this could be implemented at a later time in a object oriented interface similar to what pyfftw has? Such an interface could have other advantages.

@mreineck In any case: really great job! :)

mreineck commented:
Thanks for the analysis @ncgeib !

> I did a line profiling for 100 000 calls of a complex-valued FFT with N=1024 of the last version with cache (commit cefa941):

And this was already with an accelerated version of the caching code ... the caching in numpy 1.15 should be even slower.

mreineck commented:
I'll gladly take patches to improve error handling inside Python, since I simply don't know how to do that.

pv commented:
Controlled error handling would be easier if pocketfft functions return codes indicated error, instead of abort() on failure.

mreineck commented:
I'll see what I can do, but passing potential error codes from every function that calls malloc is probably too painful. I really miss exceptions here ... perhaps we can do something with setjmp/longjmp?

pv commented:
You can, but I don't think that's best practice in C...

mreineck commented:
I know why I'm a C++ person whenever possible ...

mreineck commented:
OK ... I think that error codes are now properly passed back to the callers in pocketfft_internal.c

However, if I really provoke an error (e.g. by returning -1 by hand from somewhere in pocketfft), it doesn't seem to be properly handled. Sometimes I get a segfault, sometimes I get
```
No traceback available to show.
Fatal Python error: GC object already tracked
Aborted
```
I have no idea what to do about this.

charris commented:
Looks like #12024 broke the merge. Should be simple enough to rebase.

mreineck commented:
> I guess that caching the twiddle-factors would only make sense in a very efficient Cython wrapper (or when caching on C-side). Maybe this could be implemented at a later time in a object oriented interface similar to what pyfftw has? Such an interface could have other advantages.

Caching on the C side is possible, but it will complicate the Python<->C interface, because we need to be able to change the caching parameters from `numpy`.
Also we would still have to hold a lock while manipulating the cache, which is a bit hairy, and we may need to do explicit cleanup operations at some point.
Not something I'm looking forward to, but maybe someone else would like to contribute this?

mreineck commented:
Travis failure seems to be due to a network timeout.

mreineck commented:
Just curious: do you think that further big changes will be requested during review (e.g. related to documentation, code style etc.)? I expect to be quite busy with other projects during the next months, so there might be some delay on my side ...

eric-wieser commented on numpy/fft/pocketfft.py:
```diff
@@ -194,12 +170,13 @@ def fft(a, n=None, axis=-1, norm=None):
-        output *= 1 / sqrt(n)
+    fct = 1.
+    if norm is not None and _unitary(norm):
+        fct = 1. / sqrt(n)
```
Better to just use `1` rather than `1.`

mreineck replied:
Will do, thanks!

eric-wieser commented:
It's not clear to me where the line between "pocketfft sources" and "numpy integration" falls. I assume you plan for pocketfft to continue to be a standalone project?

mreineck commented:
> I assume you plan for pocketfft to continue to be a standalone project?

Absolutely. But, as was mentioned above, my preferred method or error handling (exit everything it a real problem is detected) is not acceptable within the numpy context, and maybe there is more that I'm currently not aware of.

mreineck commented:
This is getting a bit off-topic, but supporting a more flexible error strategy would of course become much, much easier if there was a way of writing compiled extensions in, say, C++ (or any other language with exception support).

Are there considerations to allow this for numpy in the future?

charris commented:
> Are there considerations to allow this for numpy in the future?

It was been discussed, there was even a tentative plan to make that move but it ended with the 1.7 release kerfuffle. C++ would be a natural fit, what with smart pointers and all, and C++ is probably mature enough these days that it could be done. OTOH, writing good C++ is harder than writing C because there are so many options. I think matplotlib did that way back when and had some problems, athough I'm not really familiar with that package. @tacaswell Are you folks still using C++.

mreineck commented:
Performance seems to have degraded a bit in comparison to numpy 1.15; possibly due to the new array_function_dispatch mechanism.

eric-wieser commented:
@charris:
> it ended with the 1.7 release kerfuffle

Where can I learn about the history of that kerfuffle?

charris commented:
The azure-pipeline test failure is because of C90, so should be fixed after the 1.16 branch.

charris commented:
@tylerjereddy What is the easiest way to set the compiler flags for docker? Just set the CFLAG environment variable?

rgommers commented:
> We need to wait until we drop Python 2.7 and go to C99.

Ah yes, forgot about that. Not too much longer:)

tylerjereddy commented:
> What is the easiest way to set the compiler flags for docker? Just set the CFLAGS environment variable?

I think we can do `-e CFLAGS=something` when we create the container or even just before compiling NumPy in the container if we prefer. Sometimes it is helpful to use `NPY_DISTUTILS_APPEND_FLAGS` -- you probably know more about that than I. I used that for our coverage compiler flag adjustments in Travis on Pauli's recommendation I believe.

eric-wieser commented:
>  I have tweaked the accuracy of the twiddle factors some more, but now the code feels quite finished to me.

In just this numpy branch, or upstream in pocket fft?

If they're diverging, I think we need to make it clear that this has become a fork of pocketfft

eric-wieser commented on numpy/fft/pocketfft.c:
```diff
@@ -0,0 +1,2403 @@
+
+typedef struct cmplx {
+  double r,i;
+} cmplx;
```
If you're already committing to C99, why not use `double complex` instead?

mreineck replied:
Sure, that's possible, but the code was written at a time where one could not rely on complex support in compilers. The most important C99 feature for me was variable definitions that can occur anywhere in the code.

eric-wieser commented on numpy/fft/pocketfft.c:
```diff
@@ -0,0 +1,2403 @@
+    int fail=0;
+    Py_BEGIN_ALLOW_THREADS;
+    NPY_SIGINT_ON;
+    pl
```
I'm pretty sure that when you get to this branch, `PyErr_Occurred()` will be `NULL`, which will lead to python crashing, either with a segfault or `SystemError`

mreineck replied:
I tried to follow what the current implementation does, but I most likely missed something. I'll try to read up on error handling.

mreineck replied:
Do things look better with the recent change?

eric-wieser commented on numpy/fft/pocketfft.c:
```diff
@@ -0,0 +1,2403 @@
+    {
+    length>>=1;
+    // factor 2 should be at the front of the factor list
+    if (nfct>=NFCT) return -1;
```
It would be better to invent some error code enum, and use to distinguish the different failure types, rather than using `-1` for all failures. Then you could convert that enum into a python error at the top level.

mreineck replied:
You are right. In earlier versions the only error scenario was "out of memory", but I added this one for completeness' sake. It's still fairly academic since it can only trigger for FFT lengths larger than 2 * (3**24).

Is there a maximum length up to which a 1D FFT should be supported? If yes, maybe we should catch violations of this already in Python. One of the existing implementations (can't remember whether it was numpy of scipy) silently causes memory corruption if the FFTs become too long, which is definitely not desirable.

rgommers replied:
In terms of size, we're usually only testing on machines with up to 16 GB available; when we get bug reports for larger memory machines it's typically in the 32-128 GB range. If 128 GB is feasible (don't have a machine to test that though), that would be a good cutoff.

mreineck replied:
OK, the current choice of NFCT is so high that the (nfct>=NFCT) can never trigger on any realistic machine ... `2*3**24` is roughly `5.6e11`, and an FFT of that size would require at least 8TB of memory. I guess we can safely assume that the user will run out of memory before arriving at this check.
(In contrast, the current implementation reserves space for 15 factors, running out of space at `2*3**14`, which is around 10 million and not totally unrealistic.)

mreineck commented:
>In just this numpy branch, or upstream in pocket fft?

In both versions, so there is no need to worry about divergence.
I'm keeping both codes as similar as possible. The only difference currently is the hack to get `restrict` ignored by the Microsoft compiler.

ncgeib commented on numpy/fft/pocketfft.c:
```diff
@@ -0,0 +1,2408 @@
+  r = fma (r, s,  2.3533063028328211e-1);
+  r = fma (r, s, -1.3352627688538006e+0);
+  r = fma (r, s,  4.0587121264167623e+0);
+  r = fma (r, s, -4.9348022005446790e+0);
```
We have to watch out here on Windows platforms (and possibly others). The Visual studio compiler will not treat `fma` as an intrinsic and will call an `fma` function instead, which performs a software emulation if it not provided the correct flags [[1]](https://stackoverflow.com/questions/34265982/automatically-generate-fma-instructions-in-msvc). However this software emulation is apparently really bad and does not provide the actual fma operation with a single rounding. This is the reason fma is not available in Python yet [[2]](https://bugs.python.org/issue29282).

Unfortunately, I do not have a Windows platform at hand where I could test the accuracy of this function. But we should probably do so to make sure the fma operations are compiled correctly. Also I apologize that I brought the `sincospi` up and it now turns out to be such a hassle to implement on different platforms. Maybe we should stick with the libm implementation of sin and cos.

mreineck replied:
Also, libm `cos` wouldn't give me the `cos(x)-1` value accurately, which I need for the twiddle factors. Overall I think we will be doing well with the current implementation.

charris replied:
`cos(x) - 1 = sin(x)**2/(cos(x) + 1)` should have good accuracy for `cos(x) > 0`.

mreineck replied:
Thanks, that's very handy!
Still, the accuracy plots seem to indicate that we won't need to go there.

mreineck commented:
Accuracy of pocketfft, if `fma` is replaced by a naive implementation with double rounding:


![accuracy_complex](https://user-images.githubusercontent.com/898507/49021365-be98bc00-f192-11e8-84b0-fd6be77f43b0.png)
![accuracy_real](https://user-images.githubusercontent.com/898507/49021366-bf315280-f192-11e8-9eb4-b0ae0ca5a61e.png)

mreineck commented:
Accuracy with proper `fma`:
![accuracy_complex](https://user-images.githubusercontent.com/898507/49021552-2b13bb00-f193-11e8-8954-ca6780ae7427.png)
![accuracy_real](https://user-images.githubusercontent.com/898507/49021554-2bac5180-f193-11e8-8079-c68806d4eb33.png)

mreineck commented:
Performance of the current pocketfft version:
![speed_complex](https://user-images.githubusercontent.com/898507/49021846-da509200-f193-11e8-851c-2cdd86d0ba1d.png)
![speed_real](https://user-images.githubusercontent.com/898507/49021848-dae92880-f193-11e8-9682-6c048b19a84f.png)

eric-wieser commented on numpy/fft/pocketfft.c:
```diff
@@ -2217,8 +2217,11 @@ execute_complex(PyObject *a1, int is_forward, double fct)
-    if (fail)
-      { Py_XDECREF(data); return NULL; }
+    if (fail) {
+      PyErr_SetString(ErrorObject, "PocketFFT: memory allocation error");
```
`PyErr_NoMemory` would be more typical here

mreineck commented:
Now that 1.16 has branched, I think we should be ready to merge.

mreineck commented:
Given that this has waited for 4 months and fixes some significant, long-standing issues with the numpy FFT, I'd prefer it to go in as soon as possible. I have no idea about the time scale for the runtime selection patch, but it sounds like something that needs quite a bit of discussion and iterating over proposals.
(And since I'm not that good ad Python, I'd prefer not to have to adjust pocketfft during that process ;)

charris commented:
I was planning to merge this today or tomorrow. I think there is little to be gained by waiting back end selection, that is a separate project, and it would be good to have this well tested before the 1.17 release.

charris commented:
The 32 bit linux docker test on azure needs to be upgraded to C99.

charris commented:
Looks like we need to modify lines 344-352 in runtests.py to remove `-Werror=declaration-after-statement` and add `-C99`. I'll make a PR.

charris commented:
See  #12610.

mhvk commented:
This all looks great, but does need a release note!

Also, having not looked in any detail: the old `fft` could not do single precision; can `pocketfft`? Important for large but noisy data sets. (But obviously no reason to hold up the PR; if it were possible to implement easily, however, maybe raise an issue so it can be done later.)

charris commented:
@mreineck I've made a separate PR for the `travis-test.sh` testing fix, after that goes in you will want to rebase, probably as part of adding a release note.

tacaswell commented:
> I think matplotlib did that way back when and had some problems, athough I'm not really familiar with that package. @tacaswell Are you folks still using C++.

Yes, we still have a bunch of c++ (mostly because Agg is templated c++).

charris commented:
@mreineck Might also want to squash the 84 commits :)

mreineck commented:
> @mreineck I've made a separate PR for the `travis-test.sh` testing fix, after that goes in you will want to rebase, probably as part of adding a release note.

What would that look like? Some addition in `doc/changelog/1.17.0-changelog.rst`?

I'll fetch the latest changes from master and squash as soon as the tester runs through.

charris commented:
@mreineck Put the notes  `doc/release/1.17.0-notes.rst` under "Improvements". I think a one line list entry in  "Highlights" would also be appropriate, i.e., `* Blah, blah...`.

mhvk commented:
@mreineck - with single precision, all calculations would be single precision; in the old `fftpack`, there was
```
#define DOUBLE
#ifdef DOUBLE
#define Treal double
#else
#define Treal float
#endif
```
and then all code is defined in terms of `Treal`. I think if one were to add this, one would write the code duplication using the machinery also used for `ufuncs`. But that is for another PR! (Perhaps as part of making the `fft` routines generalized ufuncs...)

mreineck committed:
replace fftpack with pocketfft

mreineck committed:
add release notes

mreineck commented:
Simply replacing `double` by `float` will have very undesirable consequences for twiddle factor computation. But I agree, that is for another time...

mreineck commented:
OK, I think I'm done.

mreineck commented:
@charris Thanks a lot for merging!

I agree that correctness testing for FFTs is not hard to do. The previously existing unit tests have been kept, and I added identity tests (real and complex) for all lengths between 1 and 512. That should catch most potential problems.

Concerning benchmarks, I think @ncgeib's work (https://github.com/numpy/numpy/pull/11888#issuecomment-419175720) would be the best start.
